cpp/src/arrow/compute/function.h

```diff
@@ -0,0 +1,203 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// NOTE: API is EXPERIMENTAL and will change without going through a
+// deprecation cycle
+
+#pragma once
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/options.h"  // IWYU pragma: keep
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+struct ValueDescr;
+
+namespace compute {
+
+/// \brief Contains the number of required arguments for the function
+struct ARROW_EXPORT FunctionArity {
+  static FunctionArity Nullary() { return FunctionArity(0, false); }
+  static FunctionArity Unary() { return FunctionArity(1, false); }
+  static FunctionArity Binary() { return FunctionArity(2, false); }
+  static FunctionArity Ternary() { return FunctionArity(3, false); }
+  static FunctionArity Varargs(int min_args = 1) { return FunctionArity(min_args, true); }
```

Comment:
nit: VarArgs

It also seems like it might be more readable for callers to have to always specify min_args (especially because it isn't clear to me why 1 is the default and not 0 or 2

Reply:
You need at least one argument spec for type checking

Reply:
I'm changing the name, but will leave the default issue for follow up work